control-plane/grpc/src/operations/volume/traits.rs

```diff
@@ -0,0 +1,908 @@
+        }
+    }
+}
+
+impl TryFrom<volume::VolumeTarget> for VolumeTarget {
+    type Error = ReplyError;
+    fn try_from(volume_target_grpc_type: volume::VolumeTarget) -> Result<Self, Self::Error> {
+        let target = VolumeTarget::new(
+            volume_target_grpc_type.node_id.into(),
+            match volume_target_grpc_type.nexus_id {
+                Some(id) => match NexusId::try_from(id) {
+                    Ok(nexusid) => nexusid,
+                    Err(_) => return Err(ReplyError::unwrap_err(ResourceKind::Volume)),
+                },
+                None => return Err(ReplyError::unwrap_err(ResourceKind::Volume)),
+            },
+            match volume_target_grpc_type.protocol {
+                Some(i) => match volume::VolumeShareProtocol::from_i32(i) {
+                    Some(protocol) => Some(protocol.into()),
+                    None => return Err(ReplyError::unwrap_err(ResourceKind::Volume)),
+                },
+                None => None,
+            },
+        );
+        Ok(target)
+    }
+}
+
+impl From<VolumeTarget> for volume::VolumeTarget {
+    fn from(target: VolumeTarget) -> Self {
+        volume::VolumeTarget {
+            node_id: target.node().to_string(),
+            nexus_id: Some(target.nexus().to_string()),
+            protocol: match target.protocol() {
+                None => None,
+                Some(protocol) => {
+                    let protocol: volume::VolumeShareProtocol = (*protocol).into();
+                    Some(protocol as i32)
+                }
+            },
+        }
+    }
+}
+
+impl From<volume::VolumeShareProtocol> for VolumeShareProtocol {
+    fn from(src: volume::VolumeShareProtocol) -> Self {
+        match src {
+            volume::VolumeShareProtocol::Nvmf => Self::Nvmf,
+            volume::VolumeShareProtocol::Iscsi => Self::Iscsi,
+        }
+    }
+}
+
+impl From<VolumeShareProtocol> for volume::VolumeShareProtocol {
+    fn from(src: VolumeShareProtocol) -> Self {
+        match src {
+            VolumeShareProtocol::Nvmf => Self::Nvmf,
+            VolumeShareProtocol::Iscsi => Self::Iscsi,
+        }
+    }
+}
+
+impl TryFrom<get_volumes_request::Filter> for Filter {
+    type Error = ReplyError;
+    fn try_from(filter: get_volumes_request::Filter) -> Result<Self, Self::Error> {
+        Ok(match filter {
+            get_volumes_request::Filter::Volume(volume_filter) => {
+                Filter::Volume(match VolumeId::try_from(volume_filter.volume_id) {
+                    Ok(volumeid) => volumeid,
+                    Err(_) => return Err(ReplyError::unwrap_err(ResourceKind::Volume)),
+                })
+            }
+        })
+    }
+}
+
+/// trait to be implemented for CreateVolume operation
+pub trait CreateVolumeInfo: Send + Sync {
+    /// uuid of the volume
+    fn uuid(&self) -> VolumeId;
+    /// size in bytes of the volume
+    fn size(&self) -> u64;
+    /// no of replicas of the volume
+    fn replicas(&self) -> u64;
```

Comment:
nit: we tend to use u8 for the number of replicas IIRC but that's fine anyway

Reply:
https://github.com/openebs/mayastor-control-plane/blob/05650c5cf673cf1433d3a68b4991674b87cfc214/common/src/types/v0/message_bus/volume.rs#L355

In the message bus type its like this, thats why I made it u64, so that I won't need to cast again.